lib/datadog/statsd/sender.rb

```diff
@@ -2,20 +2,34 @@
 
 module Datadog
   class Statsd
+    # Sender is using a companion thread to flush and pack messages
+    # in a `MessageBuffer`.
+    # The communication with this thread is done using a `Queue`.
+    # If the thread is dead, it is starting a new one to avoid having a blocked
+    # Sender with no companion thread to communicate with (most of the time, having
+    # a dead companion thread means that a fork just happened and that we are
+    # running in the child process).
     class Sender
       CLOSEABLE_QUEUES = Queue.instance_methods.include?(:close)
 
-      def initialize(message_buffer)
+      def initialize(message_buffer, logger: nil)
         @message_buffer = message_buffer
+        @logger = logger
+        @mx = Mutex.new
       end
 
       def flush(sync: false)
-        # don't try to flush if there is no message_queue instantiated
-        return unless message_queue
-
-        message_queue.push(:flush)
+        # don't try to flush if there is no message_queue instantiated or
+        # no companion thread running
+        if !message_queue || !sender_thread.alive?
+          @logger.warn { "Statsd: can't flush: no message queue or sender_thread alive" } if @logger
```

Comment:
While this message is accurate, since it's a user-visible warning, if a user sees

```
[WARN] Statsd: can't flush: no message queue or sender_thread alive
```

show up in their logs, I'm not entirely sure they'll understand what the issue is or how they can fix it.

Perhaps we can suggest that they call `#start` in the case where there's no `message_queue`? And in the case of the `sender_thread` not being alive, is there anything we can suggest?

Reply:
I'll move these two logs to the `debug` level: they should not happen and are here to help debugging the lib in case of an issue more than anything else.